In [13]:
import pandas as pd
%load_ext autoreload
%autoreload 2
import utils
import ast
import warnings
warnings.filterwarnings("ignore")

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


In [16]:
# Ruta al dataset australian_user_reviews
ruta_review = r'../data/australian_user_reviews.json'

# Se lee de cada línea del dataset
filas_review = []
with open(ruta_review, 'r', encoding='utf-8') as f:
    for line in f.readlines():
        filas_review.append(ast.literal_eval(line))

# Se convierte en dataframe
df_reviews = pd.DataFrame(filas_review)
df_reviews


user_id                                           user_url  \
0      76561197970982479  http://steamcommunity.com/profiles/76561197970...   
1                js41637               http://steamcommunity.com/id/js41637   
2              evcentric             http://steamcommunity.com/id/evcentric   
3                  doctr                 http://steamcommunity.com/id/doctr   
4              maplemage             http://steamcommunity.com/id/maplemage   
...                  ...                                                ...   
25794  76561198306599751  http://steamcommunity.com/profiles/76561198306...   
25795           Ghoustik              http://steamcommunity.com/id/Ghoustik   
25796  76561198310819422  http://steamcommunity.com/profiles/76561198310...   
25797  76561198312638244  http://steamcommunity.com/profiles/76561198312...   
25798        LydiaMorley           http://steamcommunity.com/id/LydiaMorley   

                                                 reviews  
0      [{'funny': '', 'posted': 'Posted November 5, 2...  
1      [{'funny': '', 'posted': 'Posted June 24, 2014...  
2      [{'funny': '', 'posted': 'Posted February 3.',...  
3      [{'funny': '', 'posted': 'Posted October 14, 2...  
4      [{'funny': '3 people found this review funny',...  
...                                                  ...  
25794  [{'funny': '', 'posted': 'Posted May 31.', 'la...  
25795  [{'funny': '', 'posted': 'Posted June 17.', 'l...  
25796  [{'funny': '1 person found this review funny',...  
25797  [{'funny': '', 'posted': 'Posted July 21.', 'l...  
25798  [{'funny': '1 person found this review funny',...  

[25799 rows x 3 columns]

In [18]:
# Se revisan tipos de datos y existencias de nulos
utils.types_data_df(df_reviews)

nombre_campo        tipo_datos  no_nulos_%  nulos_%  nulos
0      user_id   [<class 'str'>]       100.0      0.0      0
1     user_url   [<class 'str'>]       100.0      0.0      0
2      reviews  [<class 'list'>]       100.0      0.0      0

In [22]:
#Se analizan duplicados desde la columna del user_id.
duplicados = df_reviews.duplicated(subset=['user_id'], keep=False)

# Filtro
duplicados_df = df_reviews[duplicados]

# Muestra el DataFrame con los duplicados
duplicados_df


user_id  \
9         76561198156664158   
50                   Rivtex   
83        76561198094224872   
119    DieMadchenschanderin   
147               relesprit   
...                     ...   
17819     76561198076474887   
17916           yolofaceguy   
18028     76561198075591109   
18234     76561198092022514   
18309        SuchGayMuchWow   

                                                user_url  \
9      http://steamcommunity.com/profiles/76561198156...   
50                   http://steamcommunity.com/id/Rivtex   
83     http://steamcommunity.com/profiles/76561198094...   
119    http://steamcommunity.com/id/DieMadchenschanderin   
147               http://steamcommunity.com/id/relesprit   
...                                                  ...   
17819  http://steamcommunity.com/profiles/76561198076...   
17916           http://steamcommunity.com/id/yolofaceguy   
18028  http://steamcommunity.com/profiles/76561198075...   
18234  http://steamcommunity.com/profiles/76561198092...   
18309        http://steamcommunity.com/id/SuchGayMuchWow   

                                                 reviews  
9      [{'funny': '', 'posted': 'Posted June 16.', 'l...  
50     [{'funny': '', 'posted': 'Posted December 23, ...  
83                                                    []  
119    [{'funny': '', 'posted': 'Posted August 29, 20...  
147    [{'funny': '', 'posted': 'Posted December 27, ...  
...                                                  ...  
17819  [{'funny': '', 'posted': 'Posted April 12.', '...  
17916  [{'funny': '', 'posted': 'Posted October 31, 2...  
18028  [{'funny': '', 'posted': 'Posted December 26, ...  
18234  [{'funny': '', 'posted': 'Posted July 3.', 'la...  
18309  [{'funny': '', 'posted': 'Posted July 18, 2015...  

[623 rows x 3 columns]

In [25]:
columna= 'user_id'
duplicados = df_reviews[df_reviews.duplicated(subset=columna, keep=False)]
if not duplicados.empty:
    # Si hay duplicados, ordena las filas duplicadas para una comparación más sencilla
    duplicados_ordenados = duplicados.sort_values(by=columna)
    print(duplicados_ordenados)
else:
    print("No hay duplicados")

                     user_id  \
12888               05041129   
5250                05041129   
3133   111222333444555666888   
3134   111222333444555666888   
4139                   29123   
...                      ...   
2721          xXAussieRockXx   
2680             yolofaceguy   
17916            yolofaceguy   
5855               zeroblade   
13975              zeroblade   

                                                user_url  \
12888              http://steamcommunity.com/id/05041129   
5250               http://steamcommunity.com/id/05041129   
3133   http://steamcommunity.com/id/11122233344455566...   
3134   http://steamcommunity.com/id/11122233344455566...   
4139                  http://steamcommunity.com/id/29123   
...                                                  ...   
2721         http://steamcommunity.com/id/xXAussieRockXx   
2680            http://steamcommunity.com/id/yolofaceguy   
17916           http://steamcommunity.com/id/yolofaceguy   
5855           

In [27]:
#verifico un user_id de ejemplo
user_id = '29123'
user_reviews = duplicados[duplicados['user_id'] == user_id]['reviews']

for review_list in user_reviews:
    for review in review_list:
        print(review['review'])
    print('-' * 40)

Can't play after the updates, people who doesn't know how to play the game, after playing the game crashes and The ♥♥♥♥ing black screen
5/10 lots of bugs and bad servers.BTW this game needs a better training facility
it needs blood when get hit, and some others changes, 9/10
What can i say? Nice graphics and nice story and charactersThe only thing wrong: it should be playable for all graphics cards even the lowest ones like 12 fps or 6But nvm nice game 8/10
Well, this game is the best i ever played but when i was downloading this game (again) i was like in 30%, but i connected today and it was 0% well if valve can fix this things i will be more happy and i will play more this game.
----------------------------------------
Can't play after the updates, people who doesn't know how to play the game, after playing the game crashes and The ♥♥♥♥ing black screen
5/10 lots of bugs and bad servers.BTW this game needs a better training facility
it needs blood when get hit, and some others change

In [29]:
# review son exactamente igual para cada registro, se borra la segunda ocurrencia.
df_reviews = df_reviews.drop_duplicates(subset='user_id', keep='first')
#verifico 
columna= 'user_id'
duplicados = df_reviews[df_reviews.duplicated(subset=columna, keep=False)]
if not duplicados.empty:
    # Si hay duplicados, ordena las filas duplicadas para una comparación más sencilla
    duplicados_ordenados = duplicados.sort_values(by=columna)
    print(duplicados_ordenados)
else:
    print("No hay duplicados")

No hay duplicados


In [30]:
# Se observa el tipo de dato que contiene 'review'
df_reviews['reviews'][0]

[{'funny': '',
  'posted': 'Posted November 5, 2011.',
  'last_edited': '',
  'item_id': '1250',
  'helpful': 'No ratings yet',
  'recommend': True,
  'review': 'Simple yet with great replayability. In my opinion does "zombie" hordes and team work better than left 4 dead plus has a global leveling system. Alot of down to earth "zombie" splattering fun for the whole family. Amazed this sort of FPS is so rare.'},
 {'funny': '',
  'posted': 'Posted July 15, 2011.',
  'last_edited': '',
  'item_id': '22200',
  'helpful': 'No ratings yet',
  'recommend': True,
  'review': "It's unique and worth a playthrough."},
 {'funny': '',
  'posted': 'Posted April 21, 2011.',
  'last_edited': '',
  'item_id': '43110',
  'helpful': 'No ratings yet',
  'recommend': True,
  'review': 'Great atmosphere. The gunplay can be a bit chunky at times but at the end of the day this game is definitely worth it and I hope they do a sequel...so buy the game so I get a sequel!'}]

### Resumen:
##### Dataset de 3 columnas y 25799 filas, sin valores nulos. Las columnas son:

    user_id: es un identificador único para el usuario.
    user_url: es la url del perfil del usuario en streamcommunity.
    reviews: contiene una lista de diccionarios.
        funny: indica si alguien puso emoticón de gracioso al review.
        posted: es la fecha de posteo del review en formato Posted April 21, 2011.
        last_edited: es la fecha de la última edición.
        item_id: es el identificador único del item, es decir, del juego.
        helpful: es la estadística donde otros usuarios indican si fue útil la información.
        recommend: es un booleano que indica si el usuario recomienda o no el juego.
        review: es una sentencia string con los comentarios sobre el juego.

In [31]:
# Se transforma a columnas cada elemento de las listas
df_reviews2 = pd.json_normalize(df_reviews['reviews'])
df_reviews2.head()

0  \
0  {'funny': '', 'posted': 'Posted November 5, 20...   
1  {'funny': '', 'posted': 'Posted June 24, 2014....   
2  {'funny': '', 'posted': 'Posted February 3.', ...   
3  {'funny': '', 'posted': 'Posted October 14, 20...   
4  {'funny': '3 people found this review funny', ...   

                                                   1  \
0  {'funny': '', 'posted': 'Posted July 15, 2011....   
1  {'funny': '', 'posted': 'Posted September 8, 2...   
2  {'funny': '', 'posted': 'Posted December 4, 20...   
3  {'funny': '', 'posted': 'Posted July 28, 2012....   
4  {'funny': '1 person found this review funny', ...   

                                                   2  \
0  {'funny': '', 'posted': 'Posted April 21, 2011...   
1  {'funny': '', 'posted': 'Posted November 29, 2...   
2  {'funny': '', 'posted': 'Posted November 3, 20...   
3  {'funny': '', 'posted': 'Posted June 2, 2012.'...   
4  {'funny': '2 people found this review funny', ...   

                                                   3  \
0                                               None   
1                                               None   
2  {'funny': '', 'posted': 'Posted October 15, 20...   
3  {'funny': '', 'posted': 'Posted June 29, 2014....   
4  {'funny': '', 'posted': 'Posted July 11, 2013....   

                                                   4  \
0                                               None   
1                                               None   
2  {'funny': '', 'posted': 'Posted October 15, 20...   
3  {'funny': '', 'posted': 'Posted November 22, 2...   
4                                               None   

                                                   5     6     7     8     9  
0                                               None  None  None  None  None  
1                                               None  None  None  None  None  
2  {'funny': '', 'posted': 'Posted October 15, 20...  None  None  None  None  
3  {'funny': '', 'posted': 'Posted February 23, 2...  None  None  None  None  
4                                               None  None  None  None  None

In [32]:
# Para no perder 'user_id' y 'user_url' a las columnas separadas, se concatena
df_reviews2 = pd.concat([df_reviews[['user_id', 'user_url']], df_reviews2], axis=1)
df_reviews2.head()

user_id                                           user_url  \
0  76561197970982479  http://steamcommunity.com/profiles/76561197970...   
1            js41637               http://steamcommunity.com/id/js41637   
2          evcentric             http://steamcommunity.com/id/evcentric   
3              doctr                 http://steamcommunity.com/id/doctr   
4          maplemage             http://steamcommunity.com/id/maplemage   

                                                   0  \
0  {'funny': '', 'posted': 'Posted November 5, 20...   
1  {'funny': '', 'posted': 'Posted June 24, 2014....   
2  {'funny': '', 'posted': 'Posted February 3.', ...   
3  {'funny': '', 'posted': 'Posted October 14, 20...   
4  {'funny': '3 people found this review funny', ...   

                                                   1  \
0  {'funny': '', 'posted': 'Posted July 15, 2011....   
1  {'funny': '', 'posted': 'Posted September 8, 2...   
2  {'funny': '', 'posted': 'Posted December 4, 20...   
3  {'funny': '', 'posted': 'Posted July 28, 2012....   
4  {'funny': '1 person found this review funny', ...   

                                                   2  \
0  {'funny': '', 'posted': 'Posted April 21, 2011...   
1  {'funny': '', 'posted': 'Posted November 29, 2...   
2  {'funny': '', 'posted': 'Posted November 3, 20...   
3  {'funny': '', 'posted': 'Posted June 2, 2012.'...   
4  {'funny': '2 people found this review funny', ...   

                                                   3  \
0                                               None   
1                                               None   
2  {'funny': '', 'posted': 'Posted October 15, 20...   
3  {'funny': '', 'posted': 'Posted June 29, 2014....   
4  {'funny': '', 'posted': 'Posted July 11, 2013....   

                                                   4  \
0                                               None   
1                                               None   
2  {'funny': '', 'posted': 'Posted October 15, 20...   
3  {'funny': '', 'posted': 'Posted November 22, 2...   
4                                               None   

                                                   5     6     7     8     9  
0                                               None  None  None  None  None  
1                                               None  None  None  None  None  
2  {'funny': '', 'posted': 'Posted October 15, 20...  None  None  None  None  
3  {'funny': '', 'posted': 'Posted February 23, 2...  None  None  None  None  
4                                               None  None  None  None  None

In [33]:
#Utilizo pd.melt para transformar las columnas en filas conservando el 'user_id' y 'user_url'
df_reviews2 = pd.melt(df_reviews2, id_vars=['user_id', 'user_url'], 
                       value_vars=list(range(9)),
                       value_name='reviews')
df_reviews2.head()

user_id                                           user_url  \
0  76561197970982479  http://steamcommunity.com/profiles/76561197970...   
1            js41637               http://steamcommunity.com/id/js41637   
2          evcentric             http://steamcommunity.com/id/evcentric   
3              doctr                 http://steamcommunity.com/id/doctr   
4          maplemage             http://steamcommunity.com/id/maplemage   

  variable                                            reviews  
0        0  {'funny': '', 'posted': 'Posted November 5, 20...  
1        0  {'funny': '', 'posted': 'Posted June 24, 2014....  
2        0  {'funny': '', 'posted': 'Posted February 3.', ...  
3        0  {'funny': '', 'posted': 'Posted October 14, 20...  
4        0  {'funny': '3 people found this review funny', ...

In [34]:
#Existen None. Por ejemplo para el primer user_id
df_reviews2[df_reviews2['user_id']=='76561197970982479']

user_id                                           user_url  \
0       76561197970982479  http://steamcommunity.com/profiles/76561197970...   
25799   76561197970982479  http://steamcommunity.com/profiles/76561197970...   
51598   76561197970982479  http://steamcommunity.com/profiles/76561197970...   
77397   76561197970982479  http://steamcommunity.com/profiles/76561197970...   
103196  76561197970982479  http://steamcommunity.com/profiles/76561197970...   
128995  76561197970982479  http://steamcommunity.com/profiles/76561197970...   
154794  76561197970982479  http://steamcommunity.com/profiles/76561197970...   
180593  76561197970982479  http://steamcommunity.com/profiles/76561197970...   
206392  76561197970982479  http://steamcommunity.com/profiles/76561197970...   

       variable                                            reviews  
0             0  {'funny': '', 'posted': 'Posted November 5, 20...  
25799         1  {'funny': '', 'posted': 'Posted July 15, 2011....  
51598         2  {'funny': '', 'posted': 'Posted April 21, 2011...  
77397         3                                               None  
103196        4                                               None  
128995        5                                               None  
154794        6                                               None  
180593        7                                               None  
206392        8                                               None

In [35]:
#Elimino los registros que tienen None en 'reviews'.
# Se eliminan las filas con valor None
df_reviews2 = df_reviews2.dropna()
# Se verifica que solo queden el 'user_id' con la cantidad de diccionarios que le corresponde
df_reviews2[df_reviews2['user_id']=='76561197970982479']

user_id                                           user_url  \
0      76561197970982479  http://steamcommunity.com/profiles/76561197970...   
25799  76561197970982479  http://steamcommunity.com/profiles/76561197970...   
51598  76561197970982479  http://steamcommunity.com/profiles/76561197970...   

      variable                                            reviews  
0            0  {'funny': '', 'posted': 'Posted November 5, 20...  
25799        1  {'funny': '', 'posted': 'Posted July 15, 2011....  
51598        2  {'funny': '', 'posted': 'Posted April 21, 2011...

In [37]:
# Se separan por columnas cada una de las claves de 'reviews'
df_reviews = df_reviews2['reviews'].apply(pd.Series, dtype='object')
df_reviews = df_reviews.add_prefix('reviews_')
df_reviews.head()

reviews_funny            reviews_posted  \
0                                    Posted November 5, 2011.   
1                                       Posted June 24, 2014.   
2                                          Posted February 3.   
3                                    Posted October 14, 2013.   
4  3 people found this review funny    Posted April 15, 2014.   

  reviews_last_edited reviews_item_id  \
0                                1250   
1                              251610   
2                              248820   
3                              250320   
4                              211420   

                                   reviews_helpful  reviews_recommend  \
0                                   No ratings yet               True   
1  15 of 20 people (75%) found this review helpful               True   
2                                   No ratings yet               True   
3   2 of 2 people (100%) found this review helpful               True   
4  35 of 43 people (81%) found this review helpful               True   

                                      reviews_review  
0  Simple yet with great replayability. In my opi...  
1  I know what you think when you see this title ...  
2  A suitably punishing roguelike platformer.  Wi...  
3  This game... is so fun. The fight sequences ha...  
4                                            Git gud

In [38]:
# Se une con el 'user_id' y 'user_url'
df_reviews = pd.concat([df_reviews2[['user_id', 'user_url']], df_reviews], axis=1)
df_reviews.head()

user_id                                           user_url  \
0  76561197970982479  http://steamcommunity.com/profiles/76561197970...   
1            js41637               http://steamcommunity.com/id/js41637   
2          evcentric             http://steamcommunity.com/id/evcentric   
3              doctr                 http://steamcommunity.com/id/doctr   
4          maplemage             http://steamcommunity.com/id/maplemage   

                      reviews_funny            reviews_posted  \
0                                    Posted November 5, 2011.   
1                                       Posted June 24, 2014.   
2                                          Posted February 3.   
3                                    Posted October 14, 2013.   
4  3 people found this review funny    Posted April 15, 2014.   

  reviews_last_edited reviews_item_id  \
0                                1250   
1                              251610   
2                              248820   
3                              250320   
4                              211420   

                                   reviews_helpful  reviews_recommend  \
0                                   No ratings yet               True   
1  15 of 20 people (75%) found this review helpful               True   
2                                   No ratings yet               True   
3   2 of 2 people (100%) found this review helpful               True   
4  35 of 43 people (81%) found this review helpful               True   

                                      reviews_review  
0  Simple yet with great replayability. In my opi...  
1  I know what you think when you see this title ...  
2  A suitably punishing roguelike platformer.  Wi...  
3  This game... is so fun. The fight sequences ha...  
4                                            Git gud

In [39]:
#Se reemplaza esos espacios como valores nulos.

df_reviews.replace('', None, inplace=True)
df_reviews.head()

user_id                                           user_url  \
0  76561197970982479  http://steamcommunity.com/profiles/76561197970...   
1            js41637               http://steamcommunity.com/id/js41637   
2          evcentric             http://steamcommunity.com/id/evcentric   
3              doctr                 http://steamcommunity.com/id/doctr   
4          maplemage             http://steamcommunity.com/id/maplemage   

                      reviews_funny            reviews_posted  \
0                              None  Posted November 5, 2011.   
1                              None     Posted June 24, 2014.   
2                              None        Posted February 3.   
3                              None  Posted October 14, 2013.   
4  3 people found this review funny    Posted April 15, 2014.   

  reviews_last_edited reviews_item_id  \
0                None            1250   
1                None          251610   
2                None          248820   
3                None          250320   
4                None          211420   

                                   reviews_helpful  reviews_recommend  \
0                                   No ratings yet               True   
1  15 of 20 people (75%) found this review helpful               True   
2                                   No ratings yet               True   
3   2 of 2 people (100%) found this review helpful               True   
4  35 of 43 people (81%) found this review helpful               True   

                                      reviews_review  
0  Simple yet with great replayability. In my opi...  
1  I know what you think when you see this title ...  
2  A suitably punishing roguelike platformer.  Wi...  
3  This game... is so fun. The fight sequences ha...  
4                                            Git gud

In [40]:
#analizo nulos
utils.types_data_df(df_reviews)

nombre_campo                           tipo_datos  no_nulos_%  \
0              user_id                      [<class 'str'>]      100.00   
1             user_url                      [<class 'str'>]      100.00   
2        reviews_funny  [<class 'NoneType'>, <class 'str'>]       13.76   
3       reviews_posted                      [<class 'str'>]      100.00   
4  reviews_last_edited  [<class 'NoneType'>, <class 'str'>]       10.28   
5      reviews_item_id                      [<class 'str'>]      100.00   
6      reviews_helpful                      [<class 'str'>]      100.00   
7    reviews_recommend                     [<class 'bool'>]      100.00   
8       reviews_review  [<class 'str'>, <class 'NoneType'>]       99.95   

   nulos_%  nulos  
0     0.00      0  
1     0.00      0  
2    86.24  49498  
3     0.00      0  
4    89.72  51499  
5     0.00      0  
6     0.00      0  
7     0.00      0  
8     0.05     30

In [41]:
#Elimino las columnas 'reviews_funny' y 'reviews_last_edited' por altos valores nulos. 
# Los de reviews se considerarán como un comentario neutral.

# Se eliminan las columnas 'reviews_funny' y 'reviews_last_edited'
df_reviews = df_reviews.drop(columns=['reviews_funny', 'reviews_last_edited'])
df_reviews.columns

Index(['user_id', 'user_url', 'reviews_posted', 'reviews_item_id',
       'reviews_helpful', 'reviews_recommend', 'reviews_review'],
      dtype='object')

In [44]:
#Transformación de la columna 'reviews_posted'
# a formato YYYY-MM-DD, pero se encuentra como Posted November 5, 2011.. Por lo tanto, es necesario procesar la fecha y extraer los elementos relevantes. Se utilizará expresiones regulares para buscar y capturar los valores de año, mes y día dentro de la cadena de texto.

df_reviews['reviews_date'] = df_reviews['reviews_posted'].apply(utils.date_converter)
df_reviews['reviews_date']

0               2011-11-05
1               2014-06-24
2         Formato inválido
3               2013-10-14
4               2014-04-15
                ...       
231291          2014-08-15
231293          2014-08-02
231419          2015-07-31
231499          2015-12-20
231501    Formato inválido
Name: reviews_date, Length: 57397, dtype: object

In [45]:
#Elimino 'reviews_posted', ya que solo indica el día y el mes del posteo.

df_reviews = df_reviews.drop('reviews_posted', axis=1)
df_reviews.columns

Index(['user_id', 'user_url', 'reviews_item_id', 'reviews_helpful',
       'reviews_recommend', 'reviews_review', 'reviews_date'],
      dtype='object')

In [46]:
# Columna 'reviews_review'
# Se eliminan los nulos

df_reviews = df_reviews.dropna(subset=['reviews_review'])
# Se verifican nulos
utils.types_data_df(df_reviews)

nombre_campo        tipo_datos  no_nulos_%  nulos_%  nulos
0            user_id   [<class 'str'>]       100.0      0.0      0
1           user_url   [<class 'str'>]       100.0      0.0      0
2    reviews_item_id   [<class 'str'>]       100.0      0.0      0
3    reviews_helpful   [<class 'str'>]       100.0      0.0      0
4  reviews_recommend  [<class 'bool'>]       100.0      0.0      0
5     reviews_review   [<class 'str'>]       100.0      0.0      0
6       reviews_date   [<class 'str'>]       100.0      0.0      0

#### Guardo Archivo limpio


In [47]:
df_reviews.to_csv('../data/user_reviews_limpo.csv', index=False)

In [ ]:
df_reviews.to_parquet('../data/user_reviews_limpo.parquet')